File: CSharpVersion/src/Exercise/CircutSolution.cs
using SIIW_Genetic.Tools;
using System;
using System.Collections.Generic;

namespace SIIW_Genetic.Exercise
{
    public class CircutSolution : Solution
    {
        private int _pathCount;
        private Path[] _paths;

        public CircutSolution(int pathCount)
        {
            _pathCount = pathCount;
            _paths = new Path[pathCount];
        }

        public CircutSolution Copy()
        {
            CircutSolution copy = new CircutSolution(_pathCount);
            for (int i = 0; i < _pathCount; i++)
            {
                copy.SetPath(i, _paths[i].Copy());
            }
            return copy;
        }

        public void SetPath(int index, Path path)
        {
            if (index < 0 || index >= _pathCount)
			    throw new IndexOutOfRangeException();

            _paths[index] = path;
        }

        public Path GetPath(int index)
        {
            if (index < 0 || index >= _pathCount)
			    throw new IndexOutOfRangeException();

            return _paths[index];
        }

        public override void PrettyPrint()
        {
            for (int i = 0; i < _pathCount; i++)
            {
                Console.Write($"Path {i}: \n\t");

                foreach (Segment segment in _paths[i].segments)
                {
                    Console.Write($"{segment.direction} {segment.length}; ");
                }
                Console.Write("\n\n");
            }
        }
    }
}

File: CppVersion/src/SIIW_1/CircutSolution.cpp
#include "CircutSolution.h"
#include "../SIIW/Exception.h"

#include <iostream>

namespace SIIW_1
{
	CircutSolution::CircutSolution(int pathCount)
		: m_pathCount(pathCount), m_paths(new Path*[pathCount])
	{
	}

	CircutSolution::CircutSolution(const CircutSolution& other)
		: m_pathCount(other.m_pathCount), m_paths(new Path*[other.m_pathCount])
	{
		for (int i = 0; i < m_pathCount; i++)
		{
			m_paths[i] = new Path(*(other.m_paths[i]));
		}
	}

	CircutSolution::~CircutSolution()
	{
		for (int i = 0; i < m_pathCount; i++)
		{
			delete m_paths[i];
		}

		delete[] m_paths;
	}

	void CircutSolution::SetPath(int index, Path* path)
	{
		if (index < 0 || index >= m_pathCount)
			throw SIIW::Exception("Index out of bounds");
		
		m_paths[index] = path;
	}

	Path* CircutSolution::GetPath(int index)
	{
		if (index < 0 || index >= m_pathCount)
			throw SIIW::Exception("Index out of bounds");
		
		return m_paths[index];
	}

	void CircutSolution::PrettyPrint()
	{
		for (int i = 0; i < m_pathCount; i++)
		{
			std::cout << "Path " << i << ": " << std::endl << "\t";

			for (auto segment : m_paths[i]->segments)
			{
				if (segment->direction == Segment::UP)
					std::cout << "UP ";
				else if (segment->direction == Segment::DOWN)
					std::cout << "DOWN ";
				else if (segment->direction == Segment::LEFT)
					std::cout << "LEFT ";
				else
					std::cout << "RIGHT ";

				std::cout << segment->length << "; ";
			}
			std::cout << std::endl << std::endl;
		}
	}
}